src/kOS.Safe/Screen/ScreenBuffer.cs

```diff
@@ -70,15 +85,45 @@ protected void AddNewBufferLines( int howMany = 1)
                 buffer.Add(new char[ColumnCount]);
         }
 
-        public void SetSize(int rowCount, int columnCount)
+        public void AddResizeNotifier(ScreenBuffer.ResizeNotifier notifier)
+        {
+            if (notifyees.IndexOf(notifier) < 0)
+                notifyees.Add(notifier);
+        }
+
+        public void RemoveResizeNotifier(ScreenBuffer.ResizeNotifier notifier)
+        {
+            notifyees.Remove(notifier);
+        }
+
+        public void SetSize(int rows, int columns)
+        {
+            RowCount = rows;
+            ColumnCount = columns;
+            ResizeBuffer();
+            int scrollDiff = 0;
+            foreach (ResizeNotifier notifier in notifyees)
+            {
+                if (notifier != null)
+                    scrollDiff += notifier(this);
+            }
+            ScrollVertical(scrollDiff);
+        }
```

Comment:
The above method is where the meat of the change here comes in.  In the past you only ever initialized the ScreenBuffer once and then it stayed as it was - the same size forever.  But since the ScreenBuffer holds the entire "image" of the terminal (or at least the 2D array of its chars even its blank padding chars), it needs to match the size of the terminal.  Thus it needs to be told to resize itself when the terminal is stretched or shrunk.